crypto/thread/api.c

```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright 2019-2020 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/configuration.h>
+#if defined(OPENSSL_THREADS)
+
+# include <openssl/crypto.h>
+# include <internal/worker.h>
+# include <internal/thread.h>
+# include "openssl_threads.h"
+# include "worker.h"
+
+int CRYPTO_THREAD_enabled(OPENSSL_CTX *ctx)
+{
+    OPENSSL_CTX_THREADS tdata;
+
+    tdata = OPENSSL_CTX_THREADS(ctx);
+    if (tdata == NULL)
+        return 0;
+
+    return tdata->enabled;
+}
+
+int CRYPTO_THREAD_enable(OPENSSL_CTX *ctx, int max_threads)
+{
+    OPENSSL_CTX_THREADS tdata;
+
+    tdata = OPENSSL_CTX_THREADS(ctx);
+    if (tdata == NULL)
+        return 0;
+
+    tdata->threads.cap = max_threads;
+    tdata->enabled = 1;
+    return 1;
+}
+
+int CRYPTO_THREAD_disable(OPENSSL_CTX *ctx)
+{
+    OPENSSL_CTX_THREADS tdata;
+    CRYPTO_WORKER worker;
+    struct list *iter, *tmp;
+
+    tdata = OPENSSL_CTX_THREADS(ctx);
+    if (tdata == NULL)
+        return 0;
+
+    CRYPTO_MUTEX_lock(tdata->lock);
+    tdata->threads.cap = 0;
+    tdata->enabled = 0;
+
+    list_for_each_safe(iter, tmp, &tdata->workers.available) {
+        worker = container_of(iter, struct crypto_worker_st, list);
+        CRYPTO_THREAD_native_terminate(worker->handle);
+        CRYPTO_THREAD_native_clean(worker->handle);
+        list_del(iter);
+        OPENSSL_free(worker);
+    }
+    list_for_each_safe(iter, tmp, &tdata->workers.busy) {
+        worker = container_of(iter, struct crypto_worker_st, list);
+        CRYPTO_THREAD_native_terminate(worker->handle);
+        CRYPTO_THREAD_native_clean(worker->handle);
+        list_del(iter);
+        OPENSSL_free(worker);
+    }
+    list_for_each_safe(iter, tmp, &tdata->threads.active) {
+        worker = container_of(iter, struct crypto_worker_st, list);
+        CRYPTO_THREAD_native_terminate(worker->handle);
+        CRYPTO_THREAD_native_clean(worker->handle);
+        list_del(iter);
+        OPENSSL_free(worker);
+    }
+    CRYPTO_MUTEX_unlock(tdata->lock);
+    CRYPTO_THREAD_clean(ctx, NULL);
+    return 1;
+}
+
+int CRYPTO_THREAD_cap(OPENSSL_CTX *ctx, int max_threads)
+{
+    OPENSSL_CTX_THREADS tdata;
+
+    tdata = OPENSSL_CTX_THREADS(ctx);
+    if (tdata == NULL)
+        return 0;
+
+    if (tdata->enabled == 0)
+        return 0;
+
+    tdata->threads.cap = max_threads;
+    return 1;
+}
+
+size_t CRYPTO_THREAD_get_available_threads(OPENSSL_CTX *ctx)
```

Comment:
Perhaps "num" instead of "get"

Reply:
Resolved: https://github.com/ckalina/openssl/commit/3a23df4b93e1364ed5ae9e1802ebb9a2fd54975f